Allocation profiler

vilterp opened the pull request:
## Overview

Record the type and stack of every allocation (or only at a given sample interval), and return as Julia objects.

Alternate approach to existing alloc profiler PR: https://github.com/JuliaLang/julia/pull/33467
Complementary to garbage profiler PR: https://github.com/JuliaLang/julia/pull/42658 (maybe there's some nice way to meld them)

This may be reinventing the wheel from #33467, but I'm not sure why that one needs stuff like LLVM passes. I mimicked some stuff from it, but this was my attempt to get something up and running. Could easily be missing stuff.

## Usage:

```julia
using Profile.Allocs
res = Allocs.@profile sample_rate=0.001 my_func()
prof = Allocs.fetch()
# do something with `prof`
```

See also: https://github.com/JuliaPerf/PProf.jl/pull/46 for support for visualizing these.

Here are two sample screenshots from analyzing all allocations during function compilation and evaluation:
```julia
julia> Profile.Allocs.@profile sample_rate=1 begin
           @eval begin f() = 1; f() end
       end
1

julia> PProf.Allocs.to_pprof(Profile.Allocs.fetch(), from_c=false)
"alloc-profile.pb.gz"
```
<img width="1261" alt="Screen Shot 2022-01-07 at 11 55 58 AM" src="https://user-images.githubusercontent.com/1582097/148578838-60fedb94-bc3d-4139-b38c-7cf73baeefcf.png">
<img width="1336" alt="Screen Shot 2022-01-07 at 11 56 21 AM" src="https://user-images.githubusercontent.com/1582097/148578842-c4f0acd8-2676-44e8-b126-a989c4540c5b.png">


## Output:

Julia structs which can be used by downstream libraries. (See `PProf.jl`: https://github.com/JuliaPerf/PProf.jl/pull/46)

## Performance

Recording the stack traces with LibUnwind for each allocation has non-trivial overhead, so the allocs profiler comes with a `sample_rate = 0.0001` that controls the probability of recording a sample for each allocation. With `sample_rate = 1`, this example benchmark which allocates heavily (`for i in 1:100 fi = Symbol("f$i") ; @eval begin $fi() = 3 ; $fi() end`) spent around 50% of time recording the stack traces (on my MacBook Pro):

<img width="300" alt="Screen Shot 2022-01-06 at 2 55 05 PM" src="https://user-images.githubusercontent.com/1582097/148444290-2ed5a7b2-5b26-4dde-827e-35c22219cfe2.png">

Since almost all the overhead is in `jl_unw_stepn()`, there's not much more that can be improved in this PR. This is why we let users control the sample rate, based on their particular workload.

## TODO:
- [x] move decoding into Julia
- [x] handle multiple threads
- [x] take `record_allocated_value` frame off the top
- [x] more robust solution for detecting non-pointer type tags (i.e. remove `TYPE_PTR_LOW_THRESHOLD`)
- [x] make it into a real stdlib package (or fold into `Profile` package); ensure tests are running
- [x] docs and news
- [ ] in future PRs: capture remaining allocations (incl. ones that don't go through this codepath). filed as https://github.com/JuliaLang/julia/issues/43688

Co-Authored-By: @NHDaly

vilterp committed:
alloc profiler

vilterp committed:
fix key

vilterp committed:
don't skip stdlib

vilterp committed:
rename

vilterp committed:
record freed values

vilterp committed:
print frees by type

strangely empty

vilterp committed:
fix recording & printing of frees

NHDaly commented on base/gcutils.jl:
```diff
@@ -105,6 +105,14 @@ Control whether garbage collection is enabled using a boolean argument (`true` f
+    ccall(:jl_start_alloc_profile, Cvoid, (Cint,), skip_every)
+end
+
+function stop_and_write_alloc_profile(io)
+    ccall(:jl_stop_and_write_alloc_profile, Cvoid, (Ptr{Cvoid},), io.handle)
+end
```
```suggestion
function stop_and_write_alloc_profile(filename::String)
    open(filename, "w") do io
        stop_and_write_alloc_profile(io)
    end
end
function stop_and_write_alloc_profile(io)
    ccall(:jl_stop_and_write_alloc_profile, Cvoid, (Ptr{Cvoid},), io.handle)
end
```

NHDaly commented on src/gc-alloc-profiler.cpp:
```diff
@@ -0,0 +1,517 @@
+        jl_safe_printf("No code info - unknown interpreter state!\n");
+    }
+    return ret;
+}
```
I'm getting a segfault during `GC.stop_and_write_alloc_profile("gc-profile.json")`:
```julia
No code info - unknown interpreter state!

signal (11): Segmentation fault: 11
in expression starting at REPL[6]:1
__cxx_atomic_load<_jl_value_t *> at /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/atomic:970 [inlined]
load at /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/atomic:1487 [inlined]
atomic_load_explicit<_jl_value_t *> at /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/atomic:1828 [inlined]
jl_atomic_load_explicit<_jl_value_t *> at /Users/nathandaly/src/julia/src/./julia_atomics.h:93 [inlined]
jl_array_ptr_ref at /Users/nathandaly/src/julia/src/./julia.h:950 [inlined]
get_julia_frames at /Users/nathandaly/src/julia/src/gc-alloc-profiler.cpp:193
get_frames at /Users/nathandaly/src/julia/src/gc-alloc-profiler.cpp:266
expand_stack at /Users/nathandaly/src/julia/src/gc-alloc-profiler.cpp:299
profile_serialize at /Users/nathandaly/src/julia/src/gc-alloc-profiler.cpp:350
alloc_profile_serialize at /Users/nathandaly/src/julia/src/gc-alloc-profiler.cpp:420
jl_stop_and_write_alloc_profile at /Users/nathandaly/src/julia/src/gc-alloc-profiler.cpp:439
stop_and_write_alloc_profile at ./gcutils.jl:113 [inlined]
#1 at ./REPL[2]:3 [inlined]
#open#363 at ./io.jl:384
open at ./io.jl:382 [inlined]
stop_and_write_alloc_profile at ./REPL[2]:2 [inlined]
top-level scope at ./REPL[6]:6
jl_toplevel_eval_flex at /Users/nathandaly/src/julia/src/toplevel.c:880
jl_toplevel_eval_flex at /Users/nathandaly/src/julia/src/toplevel.c:833
jl_toplevel_eval_flex at /Users/nathandaly/src/julia/src/toplevel.c:833
ijl_toplevel_eval at /Users/nathandaly/src/julia/src/toplevel.c:898 [inlined]
ijl_toplevel_eval_in at /Users/nathandaly/src/julia/src/toplevel.c:948
```

NHDaly replied:
@maleadt does anything stand out to you here? @vilterp says he copied this code from `stackwalk.c`.

maleadt replied:
No, but I'm not too familiar with that code.

vilterp commented:
Still not sure how best to capture the `gc_pool_alloc` allocs (either conditionally emit an LLVM intrinsic, or plumb the type to it), but in the meantime, it'd be good to make this PR be more like the stdlib `Profile` module (and https://github.com/JuliaLang/julia/pull/33467/files), namely:

1. record the backtraces directly to a buffer, and do the stack trace decoding in Julia (having more logic in Julia is good in general, and hopefully this will fix the segfault we saw)
2. have an API in `Profile`, rather than `GC`; something like
```julia
Profile.@profile_allocs foo()
```

which would be defined as something like

```
macro profile_allocs(ex)
    return quote
        try
            enable_alloc_profiling()
            if status < 0
                error(error_codes[status])
            end
            $(esc(ex))
        finally
            disable_alloc_profiling()
        end
    end
end
```

NHDaly reviewed (commented):
Similar to @vilterp's comment above, I'd like to take a different approach to this PR than we took in the previous attempts, in order to avoid getting stuck as those PRs have been.

I think an effective approach to this would be to merge the allocations profiling functionality in stages. My impression is that the previous attempts at a memory profiler have gotten stuck in trying to work out the best way to capture the allocations coming from codegen (e.g. `gc_pool_alloc`). This PR currently does not capture those allocations, but it _does_ still report a meaningfully large proportion of the allocations. From our experiments (measuring against `@time`) it seems to capture roughly 80% of the allocations in our heavy benchmarks, so these seems still quite useful as-is.

I'd like us to wrap up the existing functionality (profiling all allocations that go through `jl_gc_alloc()`), and get the PR into a good state, and then review it and merge it, even though it doesn't capture _all_ allocations.

We can file an issue describing the fact that the memory profiler doesn't capture all allocations, with some suggested work plans, and then follow-up with more PRs later. I think that if we add some clear documentation, and maybe even print a warning to the console that "This feature is experimental, and currently does not capture all allocations. The reported allocations are accurate, but maybe be missing a large proportion of the total allocations.", we could merge this and still provide some value, without letting this PR bit rot! :) 

Does that sound right to others? @vilterp and I will work this week to wrap up the existing functionality and get things tested and cleaned up. :)

NHDaly committed:
Return the raw data from the C++ allocations vector into julia!! :)

vilterp committed:
get started on decoding on Julia side

vilterp committed:
delete more serialization code

vilterp committed:
delete destructors and constructors

gonna try unique_ptr

vilterp committed:
use unique_ptr

not sure it's actually freeing, but at least it doesn't crash

vilterp committed:
separate free function; move Julia decoding to stdlib

vilterp committed:
move more to the stdlib

vilterp committed:
start hacking on _reformat_bt

adapted from Tim's PR

vilterp committed:
fix decoding

vilterp committed:
get alloc size and type address

next up: load actual type

vilterp committed:
just keep type type address for now

vilterp committed:
send type names over to the Julia side

kind of annoying, but it doesn't seem we can reliably load them from
the addresses without segfaulting

vilterp committed:
package up frees as well

vilterp committed:
try to fix

but I think the problem is that the strings are being GC'd

cuz they're not rooted

vilterp committed:
malloc strings so they're not junk by the time we look at them

this works, but all the strings leak
ugh C is hard

vilterp committed:
implement caching of stack trace lookups

like the Profile does, and the C++ side did before

wonder if we can move this into stacktraces.jl or something to
dedup with Profile

vilterp committed:
add GC.@preserve, magically fixing segfaults (???)

also change Julia representation of raw backtrace elements,
thogh I'm not sure it was breaking anything

vilterp committed:
delete code to stringify type names on C++ side

thanks to GC.@preserve <3

NHDaly committed:
Fix hang in GC: use `jl_safe_printf` not `jl_printf`

NHDaly committed:
Mark the AllocProfile C code as `JL_NOTSAFEPOINT`

TODO: is this correct!?

vilterp commented:
Sometimes get

```
ERROR: TypeError: in typeassert, expected Union{Nothing, Core.CodeInfo, Core.MethodInstance}, got a value of type Vector{UInt8}
Stacktrace:
  [1] _reformat_bt(bt::Vector{Ptr{Nothing}}, bt2::Vector{Any})
    @ Base ./error.jl:88
  [2] decode_alloc(cache::Dict{Union{Ptr{Nothing}, Base.InterpreterIP}, Vector{Base.StackTraces.StackFrame}}, raw_alloc::Main.AllocProfile.RawAlloc)
    @ Main.AllocProfile ~/code/julia/stdlib/AllocProfile/src/AllocProfile.jl:94

```

with this strategy

vilterp commented:
EDIT: yeah, things are getting GC'd out from under us — if I do `GC.enable(false)`, these type errors go away 🤦‍♂️

vilterp commented:
Weird thing is, it works sometimes, even when full GCs happen during the collection and decode phases:
```
julia> AllocProfile.start(100)

julia> @time HeapSnapshotParser.parse_snapshot("testdata/empty-repl.heapsnapshot")
GC: pause 68.566574ms. collected 213.528655MB. 229376366 allocs total. incr 
GC: pause 67.191154ms. collected 39.250776MB. 45875220 allocs total. incr 
GC: pause 41.633240ms. collected 44.589300MB. 45875218 allocs total. incr 
GC: pause 52.344458ms. collected 42.013328MB. 45875207 allocs total. full 
GC: pause 105.485016ms. collected 79.103016MB. 91751390 allocs total. full 
GC: pause 86.088875ms. collected 43.294896MB. 45875392 allocs total. incr 
GC: pause 21.218635ms. collected 42.798672MB. 45875768 allocs total. incr 
GC: pause 20.006157ms. collected 44.481792MB. 45876888 allocs total. incr 
GC: pause 22.843812ms. collected 44.484160MB. 45876912 allocs total. incr 
GC: pause 22.722307ms. collected 41.068672MB. 45876632 allocs total. incr 
GC: pause 24.197558ms. collected 37.902328MB. 45875215 allocs total. incr 
GC: pause 20.181365ms. collected 40.804064MB. 45875208 allocs total. incr 
  1.416460 seconds (4.32 M allocations: 372.909 MiB, 22.79% gc time, 13.22% compilation time)
HeapSnapshot

julia> results = @time AllocProfile.stop()
GC: pause 29.494241ms. collected 38.331040MB. 45876416 allocs total. incr 
GC: pause 10.257597ms. collected 30.266950MB. 45875204 allocs total. incr 
GC: pause 9.489497ms. collected 21.904032MB. 46197488 allocs total. incr 
GC: pause 24.214812ms. collected 25.502704MB. 53997360 allocs total. full 
GC: pause 85.942493ms. collected 55.266176MB. 54010624 allocs total. full 
GC: pause 82.195946ms. collected 26.697176MB. 54003344 allocs total. full 
GC: pause 91.935260ms. collected 67.363824MB. 134991008 allocs total. full 
GC: pause 115.030000ms. collected 156.418160MB. 337481136 allocs total. full 
GC: pause 177.543380ms. collected 340.834528MB. 843697408 allocs total. full 
  3.283125 seconds (15.71 M allocations: 1.746 GiB, 18.17% gc time, 31.47% compilation time)
AllocResults
```

vilterp committed:
print incr. vs full

vilterp committed:
simpler Base.show for AllocResults

timholy commented:
I haven't looked at the code, but it seems like you'd want to record the data at the moment of allocation, in which case

> things are getting GC'd out from under us 

doesn't seem like it would be a concern. But again, I haven't looked at this due to time constraints..

vilterp commented:
Thanks Tim. We are recording backtraces at the point of allocation. When the backtraces are solely composed of instruction pointers, everything seems to work fine. But when they also include pointers to objects on the heap, it seems those objects are sometimes garbage collected before we decode the stack traces.

We think these are objects associated with the interpretation (as opposed to JIT) mode. Stack frames for interpreted functions leave pointers in the stack trace to objects like `CodeInfo`. I believe the problem occurs when these objects are GC'd.

timholy commented:
:+1: You should be able to serialize those in some more permanent form at the time of collection, possibly under a `GC.@preserve` or something.

vilterp commented:
Yeah, it does seem like we either need to either:
1. keep all these objects rooted til decoding time, or
2. decode/serialize more during the recording phase, so we don't have to worry about them.
2 seems nice, but we were hoping to get away with keeping most of the decoding logic in Julia at the end, so we can use functions in `stacktraces.jl` and `error.jl` like `_reformat_bt`, `lookup`, etc. Maybe doing a bit of decoding in C++ isn't too bad though 🤔 I'll take a look tomorrow.

NHDaly commented:
Ah, bummer! Makes sense though! Thanks @timholy for chiming in. That does make sense now that you explain it, @vilterp and @timholy. thanks!

Yeah, i'd be probably fine with either approach.

For `1.`, we were hoping to avoid allocating any GC-tracked objects during profiling, since this is itself a GC allocations profiler, which is why we didn't want to create any julia-managed objects to point at them...

You could maybe address that by pre-allocating a global Julia vector, @vilterp, and then stuffing the things you need to track into those? That way there's still no allocations? The only weirdness is that `Base.gc_live_bytes()` would technically still go up, since the vector would may end up growing capacity, but i doubt that anyone will be tracking things _that_ precisely? 🤔 

Doing option `2.` could work too, yeah. Very interesting (and annoying) dilemma. 👍  Thanks!

vilterp commented on stdlib/AllocProfile/src/AllocProfile.jl:
```diff
@@ -0,0 +1,189 @@
+    return unsafe_pointer_to_objref(ptr)
+end
+
+# same as _reformat_bt, except has a more specific type for bt2
```
this makes it avoid an illegal instruction error:

```
signal (4): Illegal instruction: 4
in expression starting at REPL[4]:1
decode at /Users/vilterp/code/julia/stdlib/AllocProfile/src/AllocProfile.jl:114
unknown function (ip: 0x149eba06e)
_jl_invoke at /Users/vilterp/code/julia/src/gf.c:0 [inlined]
ijl_apply_generic at /Users/vilterp/code/julia/src/gf.c:2451
jl_apply at /Users/vilterp/code/julia/src/./julia.h:1773 [inlined]
do_call at /Users/vilterp/code/julia/src/interpreter.c:126
eval_body at /Users/vilterp/code/julia/src/interpreter.c:467
```

not sure why.

Tried making the same edit in the stdlib, but it caused a method error while building the stdlib. Will try again tomorrow to find a type that works for both.

vilterp commented:
Update, since calling `_reformat_bt(bt, bt2)` has caused so much weirdness (illegal instruction errors, nonsensical type errors, etc) we're going to mimic the CPU profiler and just skip `_reformat_bt` altogether, instead just calling `lookup(::Ptr{CVoid})`, which appears to return a special "unknown" stack frame for the extended entries containing interpreter frames.

If it's good enough for the CPU profiler, it's good enough for us! Also, since we don't care about keeping these interpreter frame objects rooted in Julia GC anymore, we can move the buffers back from Julia arrays into C++, which allows the allocation profiler to not allocate anymore. This seems like a sensible thing.

We'll still decode the buffers in a Julia stdlib module and return Julia structs, which can then be used by downstream libraries like PProf, or e.g. a simple JSON serializer.

vilterp committed:
put test back in; get it passing

vilterp committed:
remove gc.preserve

vilterp committed:
add top level `profile` function

vilterp committed:
clear after decoding

vilterp committed:
free memory allocated for results

vilterp committed:
first hack at multithreaded profiler

vilterp committed:
tweaks

vilterp committed:
hold onto vectors

d'oh

vilterp committed:
fix some compile errors

NHDaly committed:
Remove file println debugging

NHDaly committed:
Add precompile statements to AllocProfile package

This _drastically_ speeds up the tests, for reasons I don't exactly
understand.. I wonder if it was messing up some heuristics and deciding
to interpret the code instead of compiling it, and had some weird
corneer cases in the interpreted code or something? I dunno!

But anyway, this drastically speeds it up, so 🤷 sounds like not our
problem 😊

NHDaly committed:
add comment about precompilation

NHDaly committed:
Malloc right-sized buffers for backtraces.

Instead of allocating a maximum-sized buffer for each backtrace, we keep
a single max-sized buffer as a scratch space, write the backtrace to it,
and then once we know the size, we allocate a right-sized buffer for the
backtrace and copy it over.

Benchmark results (measured time for profiling allocations on internal
Arroyo benchmark, with `skip_every=0`):

This only slightly improves the time to record an alloctions profile:
- Before: 275.082525 seconds
- After: 245.891006 seconds

But it drastically improves the memory usage once the profiling is
completed, according to System Activity Monitor:
- Before: 17.35 GB
- After: 6.92 GB
- (Compared to 350 MB for the same task without profiling)

We could probably slightly improve the time overhead still furthur by
using a single big vector instead of a bunch of individual allocated
buffers, but this is probably about the best we could do in terms of
space usage. This would allow us to eliminate the redundant copying, and
would also amortize away the allocations of the buffers, both of which
should reduce the performance impact. But I'm guessing the time is
mostly dominated by just how long the stack traces are, and there's no
getting around that. At best, we could expect maybe like a 2x-3x
improvement from those changes, I think.

vilterp commented on stdlib/AllocProfile/src/AllocProfile.jl:
```diff
@@ -0,0 +1,168 @@
+    ccall(:jl_free_alloc_profile, Cvoid, ())
+end
+
+# top-level entry point
+# TODO: should probably be a macro instead of a
+# higher-order function
+function profile(f::Function, skip_every::Int=0)
+    start(skip_every)
+    res = f()
+    profile = stop()
+    clear()
+    return res, profile
+end
```
if we make this a macro, how do we return both the profile and the result of the code that's being profiled? 🤔

vchuravy replied:
```
macro profile(ex, skip_every)
   quote
      $start($skip_every)
      local res = $(esc(ex))
      local profile = $stop()
      $clear()
     (res, profile)
   end
end

vilterp replied:
done, thanks :)

is there a way to make `skip_every` be optional, but also come first… 🤔

vilterp committed:
test that we get a free

vilterp committed:
remove GC logging. splitting off into another PR

vilterp commented:
Split off GC logging in to https://github.com/JuliaLang/julia/pull/43511

vilterp committed:
remove a couple debugging things

vilterp committed:
add macro interface; use in test

vilterp committed:
fix whitespace

vilterp committed:
make skip_every a named optional argument to the macro

vilterp committed:
remove debug logs

I think these were added to find why it was slow, which was fixed by the
__precompile__, so we don't need them anymore?

vilterp committed:
add multithreaded test

vilterp committed:
move AllocProfile into Profile package

vilterp committed:
rename AllocProfile => Profile.Allocs

NHDaly commented on stdlib/AllocProfile/src/AllocProfile.jl:
```diff
@@ -0,0 +1,187 @@
+end
+
+"""
+    AllocProfile.@profile [skip_every=10_000] ex
```
I just realized the name here is a bit misleading/confusing:
We're not skipping every 10,000th element, we're skipping 10,000 elements after every sample. But as it's currently written, it sounds to me like "skip one every 10,000".

Maybe switch this to `sample_every=10_000`? I think that's unambiguous.

NHDaly replied:
Or we could switch to a rate, like `sample_rate=0.0001`? It seems like the current approach requires a tiny bit of paying close attention to flip the big numbers into smaller profiles in my head.

But on the other hand, the current format matches the `delay=0.01` field, in that bigger delays also mean smaller profiles.. So maybe this is actually best how you already have it, since it's more consistent. 🤔

NHDaly replied:
Oh, blech, but also `sample_every=0` doesn't make sense... So scratch that.

`skip_between=n` maybe? 

Also another problem with `sample_rate=0.001` is that then you need to convert this float into an integer anyway to do the skipping, which is weird.. Unless you actually do want to randomly sample via rand() instead of picking every N allocations, which might actually be better to avoid you missing any cyclical patterns if your skip rate and allocation patterns are in-phase?

For example, this would be a bummer:
```julia
Allocs.@profile skip_every=2 begin   # profile every third alloc (i think?)
    for _ in 1:1000
        a,b,c = Int[], Dict(), Float64[]
        # ...
    end
end
```
You'd only see the `Int[]` allocations and miss the other ones 😬 

So after thinking about that. I think I vote strongly in favor of taking a sample rate, and actually doing uniform random sampling with the provided sample rate. `1.0` means sample everything, `0.0` means sample nothing, and the default could be `0.0001` (sample every 1/10,000).

vilterp replied:
uniform sampling is done

vilterp committed:
skip another frame to avoid record_allocated_value

vilterp commented on stdlib/AllocProfile/src/AllocProfile.jl:
```diff
@@ -0,0 +1,187 @@
+    if TYPE_PTR_LOW_THRESHOLD < UInt(ptr)
+        return unsafe_pointer_to_objref(ptr)
+    end
+    return Missing
```
make own `UnknownAllocationType` type

vilterp committed:
load types: stop using threshold hack; just check for jl_buff_tag

vilterp committed:
switch to uniform sampling with rand()

instead of recording any N allocs

NHDaly commented on stdlib/Profile/src/Allocs.jl:
```diff
@@ -100,15 +100,19 @@ end
     end
-    return Missing
+    return unsafe_pointer_to_objref(ptr)
 end
```
Oh excellent. This whole commit looks great. Does it seem like it worked?

vilterp replied:
yeah, I think so! made it through the medium-size library I test it on (HeapSnapshotParser); still need to try on our larger codebase…

(would be nice to have some larger tests in this repo 🤔)

vilterp committed:
remove garbage profile stuff

the big dict is not worth the overhead for now

vilterp committed:
remove unused struct

vilterp committed:
first hack at fetch API

vilterp committed:
remove unused `profile` function

the macro is preferred

NHDaly commented on src/gc-alloc-profiler.h:
```diff
@@ -0,0 +1,56 @@
+};
+
+struct RawAllocResults {
+    void *allocs; // Alloc* (see gc-alloc-profiler.cpp)
```
Can we declare the RawAlloc struct here, too? I think it'd be nice to, since it will be consumed from Julia? Of course it doesn't _actually_ matter, since no C/C++ files will be sharing it, but it signals some intention that the struct definitions are "public-facing".

NHDaly replied:
Done in https://github.com/vilterp/julia/pull/10

NHDaly commented on src/gc-alloc-profiler.cpp:
```diff
@@ -0,0 +1,169 @@
+    g_alloc_profile_enabled = true;
+}
+
+extern "C" {  // Needed since the function doesn't take any arguments.
```
I'm not totally sure, but I _think_ you should put this around your entire implementation of the exported interface (so lines 72 to the end).

I think it happens to work with how you have it right now, but i'm not sure if that's defined by the standards or undefined and happens to work.. I think the correct thing is that your _implementations_ of these functions should match how they were declared (which is with `extern "C"` linkage).

NHDaly replied:
resolved in https://github.com/vilterp/julia/pull/10

NHDaly commented on src/gc-alloc-profiler.cpp:
```diff
@@ -0,0 +1,169 @@
+    auto value_address = (size_t)val;
+    auto type_address = profile.type_address_by_value_address.find(value_address);
+    if (type_address == profile.type_address_by_value_address.end()) {
+        return; // TODO: warn
```
Maybe we can emit an "unknown" type here? -- like `nullptr` or some other global const sentinal value or something, and handle it on the julia side?

vilterp replied:
removed this when removing the garbage profiler

NHDaly commented on src/gc-alloc-profiler.h:
```diff
@@ -0,0 +1,56 @@
+
+extern int g_alloc_profile_enabled;
+
+static inline void record_allocated_value(jl_value_t *val, size_t size) JL_NOTSAFEPOINT {
+    if (__unlikely(g_alloc_profile_enabled)) {
+        _record_allocated_value(val, size);
+    }
+}
+
+static inline void record_freed_value(jl_taggedvalue_t *tagged_val) JL_NOTSAFEPOINT {
+    if (__unlikely(g_alloc_profile_enabled != 0)) {
+        _record_freed_value(tagged_val);
+    }
+}
```
It might help future readers of the GC code who will encounter these function calls in the GC code to name them in some way to indicate that they're for profiling? Maybe like one of these?:
- `profile_record_allocated_value`
- `profiling_record_allocated_value`
- `maybe_record_allocated_value`
?

vilterp replied:
went with `maybe_record_alloc_to_profile`

NHDaly commented on stdlib/Profile/src/Allocs.jl:
```diff
@@ -0,0 +1,180 @@
+
+# matches RawBacktrace on the C side
+struct RawBacktrace
+    data::Ptr{Csize_t} # in C: *jl_bt_element_t
```
```suggestion
    data::Ptr{Cvoid} # in C: *jl_bt_element_t
```
I think using `void*` for opaque types is more idiomatic for a C interface.

NHDaly replied:
Ah, but i see down below you _aren't_ using this as an opaque pointer, you're actually loading from it directly. It's only working because the `jl_bt_element_t` struct *happens* to be exactly the same amount of bits as `Csize_t`:
https://github.com/JuliaLang/julia/blob/a67bb2bd37f23f2b4ace373c8ea58326034e9bab/src/julia_internal.h#L904-L909

This makes me a bit nervous. Is there some way that we can either use a different type here that's kept in sync with `_jl_bt_element_t`, or make this more explicit?

NHDaly replied:
Resolved in https://github.com/vilterp/julia/pull/10

vilterp committed:
small default sample rate

Co-authored-by: Nathan Daly <NHDaly@gmail.com>

vilterp committed:
update comment

vilterp commented on src/gc-alloc-profiler.cpp:
```diff
@@ -0,0 +1,167 @@
+extern "C" {  // Needed since the function doesn't take any arguments.
+
+JL_DLLEXPORT struct RawAllocResults jl_fetch_alloc_profile() {
+    // TODO: check that the results exist
```
maybe we should also check that the profile isn't still running… (i.e. you've called `stop`, which populates these results)

NHDaly replied:
😮 oh, good call... 🤔 i wonder what happens if you violate this in the CPU Profile?
```julia
julia> @profile for _ in 1:100000; Profile.clear(); Profile.fetch() ;end
```
🤔 doesn't seem to cause any problems... but _definitely_ weird! haha

Yeah, i think asserting that seems reasonable to me, too 👍 

It actually makes me wonder if maybe we want _both_ a (cheap) interface that simply dumps the stuff from C to Julia (equivalent of fetch) _and_ a separate decoding step that converts things to useful data? I'm not sure why the Profile.jl stuff separates those concepts... 🤔 I think what you have now is good though, and no reason to change it unless someone else comments on it.

NHDaly replied:
Okay we talked on a call, and we don't think there's anything that needs to be done here, since the scratch space to record profile is different from the combined results object where it's stored, so fetch() can't interleave badly with a running profile. :)

NHDaly committed:
Comments + formatting cleanup for C files

NHDaly committed:
Cleanups & comments & docstring

NHDaly committed:
Split out and update Profile.Allocs tests

NHDaly committed:
Move the extern "C" around the impl of external funcs for Allocs
profiler

NHDaly committed:
remove pritnln

NHDaly committed:
move structs to header

NHDaly committed:
Tried to fix compile error still failed...

NHDaly committed:
Revert move struct defs - use forward decl instead

NHDaly committed:
Rename BTElementT and add docstring example

vilterp committed:
remove unordered_map

NHDaly committed:
Fix typo after merge

NHDaly committed:
Finish comment

NHDaly committed:
fix typo

NHDaly reviewed (approved):
Thanks @vilterp, this is super awesome!! 😁 It LGTM!

We've started using it and it's been useful so far already. 😁 

@vchuravy, @vtjnash: Can you take a look when you get time? I'd like your opinion on the API and the implementation. 🎉 Exciting though. 😊 Thanks!

NHDaly commented:
Oh, and this needs Docs and News again, @vilterp. Thanks. I think there should be a nice little how-to-profile-allocs section in the docs. 👍

NHDaly committed:
Stop clearing the profiled Allocations on `start()`

NHDaly committed:
Fix tests after remove Garbage Profiler (frees)

vilterp committed:
print warning message

vchuravy commented on src/gc-alloc-profiler.cpp:
```diff
@@ -0,0 +1,132 @@
+using std::string;
+using std::vector;
+
+struct RawBacktrace {
```
this breaks with the naming pattern of the other julia source files. Since they are only used internally to this file I don't know if we care, but maybe but them into an anonymous namespace?

vilterp replied:
What would be a more idiomatic name, something like `raw_backtrace`? Happy to either do that or put in an anonymous namespace.

vilterp replied:
Implemented here: https://github.com/vilterp/julia/pull/16

lmk how it looks and I'll merge it

vilterp replied:
done

vchuravy commented on src/gc-alloc-profiler.cpp:
```diff
@@ -0,0 +1,132 @@
+    g_alloc_profile_enabled = true;
+}
+
+JL_DLLEXPORT struct RawAllocResults jl_fetch_alloc_profile() {
```
Could we just return a Julia tuple?

vilterp replied:
yeah I guess so 🤔 🤷‍♂️

NHDaly replied:
Seems reasonable to me. Is that more idiomatic? What's pushing you one way or the other?

vilterp committed:
rename to to `maybe_record_alloc_to_profile`

vilterp committed:
small test cleanups

NHDaly commented on stdlib/Profile/test/allocs.jl:
```diff
@@ -0,0 +1,57 @@
+@testset "alloc profiler works when there are multiple tasks on multiple threads" begin
+    NUM_TASKS = 1000
+
+    # TODO: is this always true in CI?
+    @test Threads.nthreads() > 1
```
I would probably just delete this from the test, and just trust that at least some of the CI is run with multiple threads :/

vilterp replied:
done

vilterp replied:
done

NHDaly committed:
Fix double-free in the stop,fetch,clear logic

NHDaly committed:
Add test for start stop fetch clear

NHDaly commented on stdlib/Profile/test/allocs.jl:
```diff
@@ -0,0 +1,57 @@
+    end
+    profile = Allocs.fetch()
+
+    # expecting at least 3 allocations per task:
+    # 1. the task
+    # 2. the vector
+    # 3. the buffer inside the vector
+    @test length(profile.allocs) >= 3*NUM_TASKS
+    println(length(profile.allocs))
+    first_alloc = profile.allocs[1]
+    @test first_alloc.size > 0
+    @test length(first_alloc.stacktrace) > 0
+    @test length(string(first_alloc.type)) > 0
+
```
I guess we could actually test for the expected types in here too? Something like
```suggestion
    # expecting at least 3 allocations per task:
    # 1. the task
    # 2. the vector
    # 3. the buffer inside the vector
    @test length(profile.allocs) >= 3*NUM_TASKS
    println(length(profile.allocs))
    first_alloc = profile.allocs[1]
    @test first_alloc.size > 0
    @test length(first_alloc.stacktrace) > 0
    @test length(string(first_alloc.type)) > 0
    @testset for type in (Task, Vector, #= buffer =#)
        @test length(filter(a->a.type <: type, profile.allocs)) >= NUM_TASKS
    end
```
Or something like that? (that passed for me when i tried it locally.)

NHDaly replied:
Done.

NHDaly committed:
Ah, fix tricky C++ism: missing a `&` caused accidental double-free

NHDaly committed:
Update stdlib/Profile/src/Allocs.jl

Capitalization.

NHDaly commented:
Also, CC: @chflood. I wasn't able to tag you as a reviewer for some reason, but I want to tag you in case you'd want to have a look at this PR, too. We're excited about the potential to get a better view into what's being allocated when and where! :)

vilterp committed:
a couple cleanups; remove multithread assertion

vilterp committed:
remove unordered_map import

vilterp commented:
Rebased on master to grab the new profile.md headings from #43511. Will take a crack at writing docs and news tomorrow.

vilterp committed:
first crack at docs

vilterp committed:
add news

NHDaly committed:
Apply suggestions from code review

Fixup some small typos and formatting changes

NHDaly commented on stdlib/Profile/src/Allocs.jl:
```diff
@@ -0,0 +1,170 @@
+@assert precompile(start, ())
+@assert precompile(stop, ())
+@assert precompile(fetch, ())
+
```
So @vilterp and I noticed something strange about the performance, though, that we'd like to understand and fix. I'm not sure if we can figure it out on our own.. @vchuravy or @vtjnash if you could lend your aid that would be great!

Locally, we observed that enabling the Allocations Profiler with `sample_rate=1` (it defaults to 0.0001 to avoid this overhead) makes this example about 10x-20x slower:
```julia
julia> @time for i in 1:100 fi = Symbol("f$i") ; @eval begin $fi() = 3 ; $fi() end ; end
  0.039658 seconds (49.11 k allocations: 3.107 MiB, 9.16% compilation time)

julia> @time Allocs.@profile sample_rate=1 for i in 1:100 fi = Symbol("f$i") ; @eval begin $fi() = 3 ; $fi() end ; end
  0.549963 seconds (49.11 k allocations: 3.107 MiB, 65.56% compilation time)
```
We were trying to understand where this change comes from, and so we took some CPU profiles of the above and compared them in PProf:
```julia
julia> @time @profile for i in 1:100 fi = Symbol("f$i") ; @eval begin $fi() = 3 ; $fi() end ; end
  0.047540 seconds (49.11 k allocations: 3.107 MiB, 11.30% compilation time)
```
```julia
julia> @time @profile Profile.Allocs.@profile sample_rate=1 for i in 1:100 fi = Symbol("f$i") ; @eval begin $fi() = 3 ; $fi() end ; end
  0.609904 seconds (49.98 k allocations: 3.160 MiB, 65.49% compilation time)
```

With the allocations profiler, about 45% of the time is spent in recording the stacktraces (in `jl_unw_stepn`), and (from what i understand) there's really nothing we can do about that. This is the same image we put in the PR description:
<img width="300" alt="148444290-2ed5a7b2-5b26-4dde-827e-35c22219cfe2" src="https://user-images.githubusercontent.com/1582097/148577086-f86b96f0-c12c-42fd-8f00-be1b0f2c9cb9.png">

**However:** We _also_ see that another 45% of the time is spent in these top-level `<unknown>` frames, which aren't there when we run without the allocations profiler. We don't know what these are, nor where they're coming from. We _think_ we remember learning that the unknown frames are likely to be frames spent in the Julia Interpreter instead of compiled frames? Do you have any idea what this could be interpreting, and why?

Here's the before/after comparison - first **without** Allocations Profiling, and then **with it enabled**:
<img width="1196" alt="Screen Shot 2022-01-06 at 3 32 56 PM" src="https://user-images.githubusercontent.com/1582097/148577388-f7e8fe47-fcfa-418f-8de2-73ce67734784.png">

<img width="1223" alt="Screen Shot 2022-01-06 at 3 33 18 PM" src="https://user-images.githubusercontent.com/1582097/148577496-544b8782-b51c-4ca6-ab67-98f1629dca23.png">

Finally, for clarity, here I've highlighted in purple our `record_allocated_value()` frame, which is our entry point to stack unwinding. You can see that this accounts for 45% of the profile after the change (or 90% of the time without that extra <unknown> thing).

------------------------------


Can you help us figure out what this is? It's not _urgent_, because with a smaller sample rate like even 0.01 or 0.001, the overhead is almost entirely absent. So I think we could merge this without identifying and fixing it, but it would be nice to address. I think there's potentially a 2x reduction in the overhead that's available here. 👍 Thanks in advance!!

NHDaly replied:
EDIT: Here's the annotated photo, which illustrates our question:

<img width="1202" alt="Screen Shot 2022-01-06 at 3 33 36 PM" src="https://user-images.githubusercontent.com/1582097/148588957-00c87788-6751-4797-b664-bc70d8a33884.png">

vchuravy replied:
No spontaneous theories, but I would use `perf` to get a second opinion :)

NHDaly replied:
yeah, that's a good idea, thanks!

vilterp committed:
PR feedback

vilterp committed:
Merge branch 'pv-alloc-profile-docs-feedback' into pv-alloc-profile-docs

vilterp committed:
Merge pull request #13 from vilterp/pv-alloc-profile-docs

alloc profiler: add docs and news

NHDaly reviewed (approved):
Awesome, thanks Pete. Looks great!

vilterp committed:
print out percentage of allocs missed

vilterp committed:
remove some debug printlns

vilterp committed:
PR feedback

Co-Authored-By: NHDaly@gmail.com

vilterp committed:
more PR feedback: `_` before global names

vilterp committed:
rename structs to match Julia style

vilterp committed:
Merge pull request #16 from vilterp/pv-alloc-profile-style-renames

alloc profiler: rename structs to match Julia C style

vilterp committed:
rename jl_raw_alloc_results_t => jl_profile_allocs_raw_results_t

vilterp committed:
remove `@show`s

vilterp committed:
Merge pull request #15 from vilterp/pv-alloc-profile-allocs-missed

alloc profile: print percentage of allocs missed

vilterp committed:
record string allocs

maleadt commented on doc/src/manual/profile.md:
```diff
@@ -336,6 +336,23 @@ and how much garbage it collects each time. This can be enabled with
 
+### Allocation Profiler
+
+The allocation profiler records the stack trace, type, and size of each
```
I think the documentation should mention that this doesn't actually record every allocation (i.e. not the ones coming from codegen-emitted calls to `jl_gc_pool_alloc`).

vilterp replied:
Good point; I'll add something.

Btw, trying to at least grab stacks for those, if not types; but don't have it right yet: https://github.com/vilterp/julia/pull/14

vilterp committed:
add string test

vilterp committed:
Merge pull request #18 from vilterp/pv-alloc-profile-record-strings

alloc profile: record string allocs

vilterp committed:
alloc profiler: avoid divide-by-zero errors (#19)

* avoid divide-by-zero errors when printing error message

NHDaly commented on doc/src/manual/profile.md:
```diff
@@ -336,6 +336,23 @@ and how much garbage it collects each time. This can be enabled with
+argument can be passed to speed it up by making it skip some allocations.
+Passing `sample_rate=1.0` will make it record everything (which is slow);
+`sample_rate=0.1` will record only 10% of the allocations (faster), etc.
+
```
```suggestion

NOTE: the current implementation of the Allocations Profiler _does not
capture all allocations._ You can read more about the missing allocations
and the plan to improve this, here: https://github.com/JuliaLang/julia/issues/43688.
Calling `Profile.Allocs.fetch()` will print a log line reporting the percentage
of missed allocations, so you can understand the accuracy of your profile.
```

@maleadt and @vilterp: Here's a suggested paragraph in the docs explaining the missing allocs. :) Please feel free to edit/suggest any changes to it

vilterp replied:
committed; thanks!

vilterp committed:
add paragraph about missed allocations

Co-authored-by: Nathan Daly <NHDaly@gmail.com>

NHDaly commented:
@vchuravy / @vtjnash / @chflood: I would like to merge this early next week if we can! It LGTM, tests have been passing, and I think it's already quite useful.

Would you like to review before we merge? In particular with a focus on the API? Thanks again for your help so far!

vilterp commented:
@maleadt:
> Did you mark the contents of the additional backtrace buffers?
Ah thanks for the tip; I'm understanding more and more of what you were doing in your PR :) We were trying to get away without complicating the GC algorithm, so we didn't try that.

At present, this profiler does the same thing the CPU profiler does: simply skips interpreter frames 😅 via this edge case here in `StackTraces.lookup`: https://github.com/JuliaLang/julia/blob/00646634c6a73998eaae3785eb78fea881c39502/base/stacktraces.jl#L109

It doesn't seem like we're missing much by skipping these frames, but it'd be nice to have them. We were hoping that since this issue affects the CPU profiler as well, and they both use `StackTraces.lookup`, there could be a fix for both in a separate PR. Filed an issue about it here: https://github.com/JuliaLang/julia/issues/43796

vchuravy commented on doc/src/manual/profile.md:
```diff
@@ -336,6 +336,28 @@ and how much garbage it collects each time. This can be enabled with
+Passing `sample_rate=1.0` will make it record everything (which is slow);
+`sample_rate=0.1` will record only 10% of the allocations (faster), etc.
+
+NOTE: the current implementation of the Allocations Profiler _does not
```
I think we commonly write that as an admonition eg. `!!! note`

vchuravy reviewed (approved):
I think you are also missing arrays, e.g. 

https://github.com/JuliaLang/julia/blob/45acb76a008b2291df3e272e33a7773f3dd79bc6/src/array.c#L141

vilterp committed:
use !!! note style in comment

vilterp committed:
only print warning when missed_percentage > 0

NHDaly committed:
Attempt to test the warning logs, but struggling to capture stderr
correctly

NHDaly committed:
Remove the tests, but at least exercise both cases...

NHDaly committed:
improve warning wording to indicate sampling

NHDaly committed:
Add type assertion for alloc profiler unit tests

NHDaly commented:
> I think you are also missing arrays, e.g.
> 
> https://github.com/JuliaLang/julia/blob/45acb76a008b2291df3e272e33a7773f3dd79bc6/src/array.c#L141

good call. I think @vilterp noticed the same thing yesterday, and tackled them in https://github.com/vilterp/julia/pull/17 👍 thanks

NHDaly commented:
Thanks again everyone! :) When the tests finish and pass, I'll go ahead and Squash+Merge! 🎉

vchuravy commented on src/julia_internal.h:
```diff
@@ -364,6 +365,9 @@ STATIC_INLINE jl_value_t *jl_gc_alloc_(jl_ptls_t ptls, size_t sz, void *ty)
     }
     jl_set_typeof(v, ty);
+
+    maybe_record_alloc_to_profile(v, sz);
```
Superflous whitespace changes

NHDaly replied:
do you think it should be nestled without any whitespace? I assume Pete was trying to keep the logging code visually separate from the code that actually does anything. Is that not idiomatic? Should we remove the extra lines?

vilterp replied:
removed the whitespace

vilterp committed:
remove superfluous whitespace

NHDaly commented:
Some of the builds are failing, but I believe they are all noise.

In the commit before the last whitespace commit, all tests passed except `julia-master` and `tester_linux64`.
The failure in `tester_linux64` job fails with an `OutOfMemoryError()`, but it _did_ run all the `Profile` package tests before failing, so I feel pretty confident that it's okay too 😊 
> ```julia
>     Profile                         |      107                      107    7m53.8s
> ```
It also has succeeded in the past: 3fc80ac

Finally, the failures look similar to the ones from the commit we're based on (a733e8afd6).

So I will merge now! Thanks for your help, everyone.

vilterp commented:
Next up: https://github.com/JuliaLang/julia/pull/43868

Keno commented:
This PR failed CI (analyzegc). Please fix ASAP :).

vilterp commented:
Error:
```
/cache/build/default-amdci5-3/julialang/julia-master/src/gc-alloc-profiler.cpp:123:56: error: Calling potential safepoint from function annotated JL_NOTSAFEPOINT [julia.GCChecker]
    auto& profile = global_profile.per_thread_profiles[jl_threadid()];
                                                       ^~~~~~~~~~~~~
```

DilumAluthge commented:
If it's going to take some time to fix `analyzegc`, we can revert this PR, and reland it later once you have the fix. That way, you don't have to rush to put the fix together.

jpsamaroo commented:
I'm not sure how accessing the current task's TID is considered a safepoint? I think it might need a `JL_NOTSAFEPOINT` annotation?

DilumAluthge commented:
> If it's going to take some time to fix `analyzegc`, we can revert this PR, and reland it later once you have the fix. That way, you don't have to rush to put the fix together.

https://github.com/JuliaLang/julia/pull/43872

Keno commented:
> I'm not sure how accessing the current task's TID is considered a safepoint? I think it might need a `JL_NOTSAFEPOINT` annotation?

We follow a "guilty until declared innocent" approach here :).

vtjnash commented:
You generally want jl_current_task->tid anyways

vilterp commented:
Thanks @vtjnash, trying that here: https://github.com/JuliaLang/julia/pull/43875

NHDaly commented:
Ah, thanks, it was the extra failure in the buildkite job. Sorry I didn't do a good enough job comparing that with the existing failures. Thanks for catching this, @Keno! 🙏

vilterp commented:
k, the fix has passed `analyzegc`; waiting for the rest of CI

vtjnash commented on src/gc-alloc-profiler.cpp:
```diff
@@ -0,0 +1,139 @@
+
+jl_raw_backtrace_t get_raw_backtrace() {
+    // A single large buffer to record backtraces onto
+    static jl_bt_element_t static_bt_data[JL_MAX_BT_SIZE];
```
This is not thread safe. You should be (re)using the existing buffer in the ptls to be safe here.

vilterp replied:
Whoops. Yeah, we need a buffer per thread…

Did you mean to (re)use this buffer? https://github.com/vilterp/julia/blob/c12aca890a8ae387d11db0b54351f8b61305c00b/src/julia_threads.h#L247

It seems like we should add our own (here or as a global) to avoid colliding with that… We've shied away from adding stuff to ptls to avoid breaking stuff or adding overhead, but maybe it's better than globals?

Funny our multithreaded test didn't catch this; maybe the stack traces were just garbage.

vtjnash replied:
As long as you avoid re-using this buffer in the interval between an error filling this buffer, and then allocating the exception stack, I think you should be okay here.

vtjnash replied:
The backtrace would be essentially the same, just a bit longer perhaps

NHDaly replied:
Thank you @vtjnash for pointing this out! 😅 haha it feels pretty silly that we missed that. Thanks for the help!!

The suggestion to use a buffer in ptls makes sense. I agree with @vilterp that sharing that buffer makes me slightly nervous, but reading through your description that sounds totally fine 👍 cool, thank you!

vtjnash replied:
Profiling using a different buffer entirely

NHDaly replied:
perfect, thanks.

NHDaly replied:
One more question: is it okay to write over this buffer, since it seems like it's used to scan for roots?:
https://github.com/JuliaLang/julia/blob/7ccc83ebaf0966c5782fe1d52a431ba51cc35472/src/task.c#L344-L345

https://github.com/JuliaLang/julia/blob/7ccc83ebaf0966c5782fe1d52a431ba51cc35472/src/task.c#L618-L623

I think that it's fine, from what i can read, but just want to double check one more time.

Also, i've opened a PR for this, here: https://github.com/JuliaLang/julia/pull/44114
